### PR TITLE
TypeSpec definitions for Azure Quantum service data-plane

### LINKED ISSUE #28140: [Azure Quantum (Service) - Azure Quantum] API Review
New API Review meeting has been requested.

**Service Name**: Azure Quantum (Service) - Azure Quantum
**Review Created By**: XField _
**Review Date**: 03/13/2024 09:00 AM PT
**Release Plan**: [1171](https://web.powerapps.com/apps/821ab569-ae60-420d-8264-d7b5d5ef734c?release-plan-id=ad64f80e-6d9f-ee11-be37-000d3a341a24)
**PR**: https://github.com/Azure/azure-rest-api-specs/pull/28065
**Hero Scenarios Link**: Not Provided
**Core Concepts Doc Link**: Not Provided

**Description**: Finally, TypeSpec definitions for Azure Quantum service data-plane!

It took a while but we were able to create a TypeSpec for the Quantum data-plane from the scratch, with good code organization and applying all guidelines (which are great, by the way) and just having a handful of needed suppressions.

Detailed meeting information and documents provided can be accessed [here](https://web.powerapps.com/apps/e41c69de-ef96-44ba-b287-0e012e7c13ae?view=edit&event=0b73c0b4-cddc-ee11-904d-000d3a341a24)
For more information that will help prepare you for this review, the requirements, and office hours, visit the documentation [here](https://aka.ms/azsdk/onboarding/restapischedule)

### PR DESCRIPTION
**Pull request by vxfield**
# Data Plane API - Pull Request

This PR introduces *TypeSpec* definitions for the **Azure Quantum** *data-plane REST API*.

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*API Spec engagement record*](https://aka.ms/azsdk/onboarding/restapischedule).

* Link to API Spec engagement record issue: https://github.com/Azure/azure-rest-api-specs/issues/28140

Is this review for (select one):

- **[x] a public preview**

### Change Scope

* Design Document: `N/A`
* Previous API Spec Doc: `N/A`
* Updated paths: `N/A`

**List of changes:** 

* There are no breaking changes in the actual surface-area/implementation of the **Azure Quantum REST API**. But as we created the *TypeSpec* definitions we:
  * Fixed (removed or added) `required` parameters/properties and `x-ms-mutability` via the `@visibility` decorator.
    * This may be seen as a breaking change in the client, but the service was already enforcing it anyway.
  * Added missing value constraints (that were already enforced in the service) with `@minLength`, `@maxLength`, and `@pattern` decorators.
    * This may be seen as a breaking change in the client, but the service was already enforcing it anyway.
  * Changed some `operationIds` (via auto-generated operation ids) to conform to **Azure REST API Specs guidelines**
    * In general, applying the guidelines made the `operationId` naming better to understand!
    * This will create a breaking change in the generated *protocol-layer* SDKs, but we don't have any customers that are actively using them. All of our customers use either our *Jupyter Notebook Portal* experience, the *Azure CLI*, and mostly our *convenient* *Python SDK*. We will update the *Azure CLI* and *Python SDK* wrappers/convenience-layer to adjust to these changes.
  * Used `Azure.Core` types and operations, which caused some changes in the model (for example, the name of the `body` parameter or former custom paged-list-response types by using `ResourceList<Resource> -> Paged<Resource>`)
  * Changed some models names/refs to improve the organization/modularity of the *TypeSpec* code.
    * As with the item above, we will update our convenience-layers accordingly. 
  * Applied `suppressions` only when they were absolutely needed.

**Important:** In general, the breaking changes will not be visible or affect our customers, and it's better to do them now and align with the guidelines while we are still under public preview and have no users using the generated SDKs directly.

### Viewing API changes

**Note:** The [TypeSpec API view](https://apiview.dev/Assemblies/Review/50c4480dd09f466e9fef29f4223d06f1?revisionId=3ec8c81e57e3493885f3ad096479a657) is not showing the complete *TypeSpec*, maybe a bug because the way we modularized our code.
It might be better to review the *TypeSpec* by checking out the `vxfield-quantum-Microsoft.Quantum-2024-03-01-preview` branch or via the PR's [Files changed](https://github.com/Azure/azure-rest-api-specs/pull/28065/files).

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

### Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://aka.ms/pr-suppressions) 
to get approval.

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling

 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://github.com/Azure/azure-api-style-guide/blob/main/README.md)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications

 * [Azure REST API Guidelines](https://aka.ms/azapi/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/azapi/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links

 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

<details>
  <summary>Checks stuck in `queued` state?</summary>
If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.  
</details>

fix https://github.com/Azure/azure-rest-api-specs/issues/28140

### CODE REVIEW

File: specification/quantum/Quantum.Workspace/common/base-operations.tsp [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  @doc("Operation to patch a resource.")
+  @updatesResource(Resource)
+  @patch
+  @Azure.Core.Foundations.Private.ensureVerb("PatchOperation", "PATCH")
+  op PatchOperation<Resource extends TypeSpec.Reflection.Model> is Azure.Core.Foundations.ResourceOperation<
+    Resource,
+    {
+      @doc("This request has a JSON Merge Patch body.")
+      @TypeSpec.Http.header("Content-Type")
+      contentType: "application/merge-patch+json";
```
Comment (vxfield) [inline specification/quantum/Quantum.Workspace/common/base-operations.tsp:51]:
```suggestion
      contentType: "application/json-patch+json";
```

File: specification/quantum/Quantum.Workspace/common/base-operations.tsp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  >;
+
+  @doc("Operation to create a resource.")
+  op CreateOperation<TResource extends TypeSpec.Reflection.Model> is AzureCoreOperations.ResourceCreateOrReplace<TResource>;
+
+  @doc("Operation to patch a resource.")
+  @updatesResource(Resource)
+  @patch
+  @Azure.Core.Foundations.Private.ensureVerb("PatchOperation", "PATCH")
+  op PatchOperation<Resource extends TypeSpec.Reflection.Model> is Azure.Core.Foundations.ResourceOperation<
```
Comment (vxfield) [inline specification/quantum/Quantum.Workspace/common/base-operations.tsp:46]: Check with @allenjzhang if there is a better way to reuse the Azure.Core Patch operation.

File: specification/quantum/Quantum.Workspace/operations/storage.tsp [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+@useDependency(DefaultAzureCoreVersion)
+namespace Quantum.Workspace.Operations {
+  using Quantum.Workspace.BaseOperations;
+  @autoRoute
+  interface StorageOperations {
+    #suppress "@azure-tools/typespec-azure-core/use-standard-operations" "There is no standard operation that fits this scenario"
+    @doc("Gets a URL with SAS token for a container/blob in the storage account associated with the workspace. The SAS URL can be used to upload job input and/or download job output.")
+    @segment("storage/sasUri")
+    getSasUri is PostOperation<BlobDetailsParam, SasUriResponse>;
```
Comment (vxfield) [inline specification/quantum/Quantum.Workspace/operations/storage.tsp:45]: Check with @allenjzhang if there is a better way to implement this operation.

File: specification/quantum/Quantum.Workspace/operations/jobs.tsp [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+@doc("Quantum computing data.")
+model QuantumComputingData {
+  @doc("The number of quantum computing items in the job.")
+  @visibility("read")
+  count: int64;
+}
+
+@doc("A job to be run in the workspace.")
+// Notes:
+// - The generated Swagger is not automatically adding the x-ms-discriminator-value
```
Comment (vxfield) [inline specification/quantum/Quantum.Workspace/operations/jobs.tsp:53]: Check with @allenjzhang why the discriminator value is not being added automatically.

File: specification/quantum/Quantum.Workspace/operations/sessions.tsp [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    @doc("Close an existing session.")
+    close is ActionOperation<SessionDetails>;
+
+    @doc("Get Session by its id.")
+    get is GetOperation<SessionDetails>;
+
+    #suppress "@azure-tools/typespec-azure-core/no-operation-id" "The nesting of Jobs under a Session makes the auto-generated ID to be Sessions_JobsList, which is not in a Noun_Verb format."
+    @doc("List jobs in a session.")
+    @operationId("SessionsJobs_List")
+    jobsList is ListOperation<SessionJobDetails>;
```
Comment (vxfield) [inline specification/quantum/Quantum.Workspace/operations/sessions.tsp:92]: Check with @allenjzhang why the child-resource list operation path is clashing with a top-level list operation path.

File: specification/quantum/Quantum.Workspace/common/base-operations.tsp (hunk 2 above)
Comment (vxfield) [reply specification/quantum/Quantum.Workspace/common/base-operations.tsp:46]: Allen, Johan and I discussed about the need for this custom operation.
Currently there is no support in *TypeSpec* for this scenario.
We agreed to open an Issue in the *TypeSpec Azure repo* (https://github.com/Azure/typespec-azure/issues/475) and accept this implementation.

File: specification/quantum/Quantum.Workspace/operations/storage.tsp (hunk 3 above)
Comment (vxfield) [reply specification/quantum/Quantum.Workspace/operations/storage.tsp:45]: Allen and I didn't have time to go thought this comment, but we agreed that it's a small detail that can be addressed in the future.

File: specification/quantum/Quantum.Workspace/common/base-operations.tsp (hunk 2 above)
Comment (allenjzhang) [reply specification/quantum/Quantum.Workspace/common/base-operations.tsp:46]: Discussed in meeting.

File: specification/quantum/Quantum.Workspace/operations/jobs.tsp (hunk 4 above)
Comment (allenjzhang) [reply specification/quantum/Quantum.Workspace/operations/jobs.tsp:53]: Discussed in meeting.

File: specification/quantum/Quantum.Workspace/operations/sessions.tsp (hunk 5 above)
Comment (vxfield) [reply specification/quantum/Quantum.Workspace/operations/sessions.tsp]: Allen and I investigated this (Johan also took a quick look).
I opened a *Azure TypeSpec Issue* (https://github.com/Azure/typespec-azure/issues/494) which contains the workaround that we applied (suggested by Allen).

File: specification/quantum/Quantum.Workspace/operations/jobs.tsp (hunk 4 above)
Comment (vxfield) [reply specification/quantum/Quantum.Workspace/operations/jobs.tsp:53]: Allen and I didn't have time to go thought this comment, but we agreed that it's a small detail that can be addressed in the future.

File: specification/quantum/Quantum.Workspace/operations/items.tsp [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+// Licensed under the MIT License. See License.txt in the project root for license information.
+// --------------------------------------------------------------------------------------------
+
+import "./operations.tsp";
+
+using TypeSpec.Rest;
+using TypeSpec.Versioning;
+
+/** A key-value pair. */
+model KeyValue<Value = string> {
```
Comment (allenjzhang) [inline specification/quantum/Quantum.Workspace/operations/items.tsp:12]: would using the built in type of `Record<string>` not work?

File: specification/quantum/Quantum.Workspace/operations/jobs.tsp [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+import "./operations.tsp";
+
+using TypeSpec.Rest;
+using TypeSpec.Versioning;
+
+using TypeSpec.Http;
+using OpenAPI;
+
+/** The type of the job. */
+enum JobType {
```
Comment (allenjzhang) [inline specification/quantum/Quantum.Workspace/operations/jobs.tsp:15]: Heads up. extensible enum will be changing to be modelled with union. Can do after this PR is in.

**Review by allenjzhang — APPROVED** [review]
Please take a look at the KeyValue vs using Record<T>. The built-in type would be recognized by emitter to turn into Dictionary/Map in SDK.

File: specification/quantum/Quantum.Workspace/operations/jobs.tsp (hunk 7 above)
Comment (vxfield) [reply specification/quantum/Quantum.Workspace/operations/jobs.tsp:15]: Sounds good. I'll do this in the next PR.

File: specification/quantum/Quantum.Workspace/operations/items.tsp (hunk 6 above)
Comment (vxfield) [reply specification/quantum/Quantum.Workspace/operations/items.tsp:12]: Sounds good. I'll do this in the next PR.

**Comment by vxfield** [comment]
All, sorry about the spam of "requested reviews".
I mistakenly rebase my branch with `master` instead of `main` and when I push it, some GitHub action requested the reviews...

**Pull request merged**